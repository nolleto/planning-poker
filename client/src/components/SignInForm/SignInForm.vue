<template>
  <form @submit.prevent="processForm">
    <form-group
      name="userOrEmail"
      title="Username or Email"
    >
      <form-input
        name="userOrEmail"
        :state="getFieldState('userOrEmail')"
        :error="getFieldError('userOrEmail')"
        v-validate="'required'"
        v-model="form.userOrEmail"
      />
    </form-group>

    <form-group
      name="password"
      title="Password"
    >
      <form-input
        name="password"
        type="password"
        :state="getFieldState('password')"
        :error="getFieldError('password')"
        v-validate="'required|min:6'"
        v-model="form.password"
      />
    </form-group>

    <form-group>
      <base-button
        variant="primary"
        :disabled="formHasErrors() || isLoading"
      >
        Sign in
      </base-button>
    </form-group>

    <progress-bar v-if="isLoading"/>
    <error-summary v-if="error">
      {{ error }}
    </error-summary>
  </form>
</template>

<script>
import validatorMixin from '@/mixins/validator'
import styles from '@/styles'

export default {
  name: 'SignInForm',

  props: {
    isLoading: {
      type: Boolean,
      required: true
    },

    error: {
      type: String
    }
  },

  mixins: [ validatorMixin ],

  data () {
    return {
      form: {
        userOrEmail: '',
        password: ''
      }
    }
  },

  computed: {
    bootstrap: () => styles.bootstrap
  },

  methods: {
    processForm () {
      this.$validator.validateAll()
        .then(valid => {
          if (valid) {
            const data = {
              userOrEmail: this.form.userOrEmail,
              password: this.form.password
            }

            this.$emit('sign-in', data)
          }
        })
    }
  },

  created () {
    this.$emit('clean-server-errors')
  }
}
</script>
