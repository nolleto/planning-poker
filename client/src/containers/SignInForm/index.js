import { connect } from 'vuex-connect'
import { CLEAN_ERRORS } from '@/store/modules/auth/mutations-types'
import { getFromModule } from '@/adapters/vuex'
import SignInForm from '@/components/SignInForm'

const fromModule = getFromModule('auth')

export default connect({
  stateToProps: {
    error: fromModule(({ errors }) => errors.signIn && errors.signIn[0]),
    isLoading: ({ isLoading }) => isLoading
  },

  mutationsToEvents: {
    'clean-server-errors': dispatch => {
      dispatch(`auth/${CLEAN_ERRORS}`, { path: 'signIn' })
    }
  },

  actionsToEvents: {
    'sign-in': (dispatch, data) => dispatch('auth/signIn', data)
  }
})('SignInFormContainer', SignInForm)
