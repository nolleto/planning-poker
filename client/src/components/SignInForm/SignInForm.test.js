import SignInForm from './index'
import { shallowMount } from '@vue/test-utils'

const defaultProps = {
  isLoading: false
}

const shallowComponent = (opts = { propsData: defaultProps }) => {
  return shallowMount(SignInForm, opts)
}

describe('SignInForm component', () => {
  let wrapper

  describe('Snapshots', () => {
    beforeAll(() => {
      wrapper = shallowComponent()
    })

    it('matches', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    describe('When is loading', () => {
      beforeAll(() => {
        wrapper.setProps({ isLoading: true })
      })

      it('matches', () => {
        expect(wrapper.element).toMatchSnapshot()
      })
    })

    describe('When is there is a error', () => {
      beforeAll(() => {
        wrapper.setProps({ isLoading: false, error: 'Login invalid!' })
      })

      it('matches', () => {
        expect(wrapper.element).toMatchSnapshot()
      })
    })
  })

  describe('Event Listeners', () => {
    beforeAll(() => {
      wrapper = shallowComponent()
    })

    describe('When form submit', () => {
      beforeAll(() => {
        jest.spyOn(wrapper.vm, 'processForm')
        wrapper.find('form').trigger('submit')
      })

      it('calls #processForm method', () => {
        expect(wrapper.vm.processForm).toHaveBeenCalled()
      })
    })
  })

  describe('Methods', () => {
    beforeAll(() => {
      wrapper = shallowComponent()
    })

    describe('#processForm', () => {
      beforeAll(() => {
        jest.spyOn(wrapper.vm, '$emit')
        wrapper.vm.userOrEmail = 'user@example.com'
        wrapper.setData({
          form: {
            userOrEmail: 'user@example.com',
            password: 123456
          }
        })
        wrapper.vm.processForm()
      })

      it('emits "sign-in" with payload', () => {
        expect(wrapper.vm.$emit).toHaveBeenCalledWith(
          'sign-in',
          { userOrEmail: 'user@example.com', password: 123456 }
        )
      })
    })
  })
})
