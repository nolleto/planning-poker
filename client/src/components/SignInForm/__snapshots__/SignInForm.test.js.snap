// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm component Snapshots When is loading matches 1`] = `
<form>
  <formgroup-stub
    name="userOrEmail"
    title="Username or Email"
  >
    <forminput-stub
      aria-invalid="false"
      aria-required="true"
      name="userOrEmail"
      value=""
    />
  </formgroup-stub>
   
  <formgroup-stub
    name="password"
    title="Password"
  >
    <forminput-stub
      aria-invalid="false"
      aria-required="true"
      name="password"
      type="password"
      value=""
    />
  </formgroup-stub>
   
  <formgroup-stub>
    <basebutton-stub
      disabled="true"
      variant="primary"
    >
      
      Sign in
    
    </basebutton-stub>
  </formgroup-stub>
   
  <progressbar-stub />
   
  <!---->
</form>
`;

exports[`SignInForm component Snapshots When is there is a error matches 1`] = `
<form>
  <formgroup-stub
    name="userOrEmail"
    title="Username or Email"
  >
    <forminput-stub
      aria-invalid="false"
      aria-required="true"
      name="userOrEmail"
      value=""
    />
  </formgroup-stub>
   
  <formgroup-stub
    name="password"
    title="Password"
  >
    <forminput-stub
      aria-invalid="false"
      aria-required="true"
      name="password"
      type="password"
      value=""
    />
  </formgroup-stub>
   
  <formgroup-stub>
    <basebutton-stub
      variant="primary"
    >
      
      Sign in
    
    </basebutton-stub>
  </formgroup-stub>
   
  <!---->
   
  <errorsummary-stub>
    
    Invalid login
  
  </errorsummary-stub>
</form>
`;

exports[`SignInForm component Snapshots matches 1`] = `
<form>
  <formgroup-stub
    name="userOrEmail"
    title="Username or Email"
  >
    <forminput-stub
      aria-invalid="false"
      aria-required="true"
      name="userOrEmail"
      value=""
    />
  </formgroup-stub>
   
  <formgroup-stub
    name="password"
    title="Password"
  >
    <forminput-stub
      aria-invalid="false"
      aria-required="true"
      name="password"
      type="password"
      value=""
    />
  </formgroup-stub>
   
  <formgroup-stub>
    <basebutton-stub
      variant="primary"
    >
      
      Sign in
    
    </basebutton-stub>
  </formgroup-stub>
   
  <!---->
   
  <!---->
</form>
`;
